banzai_nres/wavelength.py

```diff
@@ -179,29 +187,49 @@ def fit_wavelength_model(features, line_list, m0):
         wavelengths_to_fit = find_nearest(features['wavelength'], np.sort(line_list))
+        residuals = wavelengths_to_fit - features['wavelength']
         weights = np.ones_like(wavelengths_to_fit, dtype=float)
         # consider weights = features['flux']/features['flux_err'] or 1/features['flux_err']**2
         # reject lines who have residuals with the line list in excess of 0.1 angstroms (e.g. reject outliers)
+        # establish an initial solution:
         weights[~np.isclose(wavelengths_to_fit, wcs.measured_lines['wavelength'], atol=0.1)] = 0
         wcs.model_coefficients = wcs.solve(wcs.measured_lines, wavelengths_to_fit, weights)
+        # iteratively refine the wavelength solution:
+
+        """
+        # call xwavecals internal routine for iteratively refining the wavelength solution.
+        # wcs.measured_lines['weight'] = 1 # uncomment and set to weights, could do any weighting scheme.
+        wcs, residuals = refine_wcs(wcs, wcs.measured_lines, np.sort(line_list), SolutionRefineOnce._converged,
+                                    SolutionRefineOnce._clip, max_iter=20,
+                                    kwargs={'sigma': 4, 'stdfunc': robust_standard_deviation})
+
+        logger.info(f'Final robust standard deviation after'
+                    f' refining: {robust_standard_deviation(residuals)} Angstrom')
+        """
         return wcs
 
 
 class IdentifyFeatures(Stage):
     """
     Stage to identify all sharp emission-like features on an Arc lamp frame.
     """
-    nsigma = 15.0  # minimum signal to noise for the feature (sum of flux / sqrt(sum of error^2) within the aperture)
+    nsigma = 10.0  # minimum signal to noise for the feature (sum of flux / sqrt(sum of error^2) within the aperture)
     # where the aperture is a circular aperture of radius fwhm.
-    fwhm = 6.0  # fwhm estimate of the elliptical gaussian PSF for each feature
-    num_features_max = 4000  # maximum number of features to keep. Will keep the num_features_max highest S/N features.
+    fwhm = 4.0  # fwhm estimate of the elliptical gaussian PSF for each feature
+    num_features_max = 2500  # maximum number of features to keep per fiber. Will keep highest S/N features up to
+    # num_features_max.
+
+    # Note that the number of lines can affect the quality of the wavelength solution. I.e. num_features_max = 4000
+    # may let too many lines into the wavelength solving procesds that are NOT in the line list and are therefore essentially
+    # spurious features. One should take great care when changing num_features_max to make sure that the wavelength solution
+    # quality is not reduced.
 
     def do_stage(self, image):
         # identify emission feature (pixel, order) positions.
-        features = identify_features(image.data, image.uncertainty, image.mask, nsigma=self.nsigma/2, fwhm=self.fwhm)
+        features = identify_features(image.data, image.uncertainty, image.mask, nsigma=self.nsigma - 1,
```

Comment:
Why is this self.nsigma-1 rather than self.nsigma?

Reply:
added comment here to address this.